k8s/apps/replicasets.go

```diff
@@ -0,0 +1,202 @@
+type ReplicaSetOps interface {
+	// CreateReplicaSet creates the given ReplicaSet
+	CreateReplicaSet(rs *appsv1.ReplicaSet) (*appsv1.ReplicaSet, error)
+	// ListReplicaSets lists all ReplicaSets in given namespace
+	ListReplicaSets(namespace string, listOpts metav1.ListOptions) ([]appsv1.ReplicaSet, error)
+	// GetReplicaSet gets the the daemon set with given name
+	GetReplicaSet(string, string) (*appsv1.ReplicaSet, error)
+	// ValidateReplicaSet checks if the given ReplicaSet is ready within given timeout
+	ValidateReplicaSet(name, namespace string, timeout time.Duration) error
+	// GetReplicaSetPods returns list of pods for the ReplicaSet
+	GetReplicaSetPods(*appsv1.ReplicaSet) ([]corev1.Pod, error)
+	// UpdateReplicaSet updates the given daemon set and returns the updated rs
+	UpdateReplicaSet(*appsv1.ReplicaSet) (*appsv1.ReplicaSet, error)
+	// DeleteReplicaSet deletes the given ReplicaSet
+	DeleteReplicaSet(name, namespace string) error
+	// GetReplicaSetByDeployment deletes the given ReplicaSet
+	GetReplicaSetByDeployment(deployment *appsv1.Deployment) (*appsv1.ReplicaSet, error)
+}
+
+// CreateReplicaSet creates the given ReplicaSet
+func (c *Client) CreateReplicaSet(rs *appsv1.ReplicaSet) (*appsv1.ReplicaSet, error) {
+	if err := c.initClient(); err != nil {
+		return nil, err
+	}
+
+	return c.apps.ReplicaSets(rs.Namespace).Create(rs)
+}
+
+// ListReplicaSets lists all ReplicaSets in given namespace
+func (c *Client) ListReplicaSets(namespace string, listOpts metav1.ListOptions) ([]appsv1.ReplicaSet, error) {
+	if err := c.initClient(); err != nil {
+		return nil, err
+	}
+
+	rsList, err := c.apps.ReplicaSets(namespace).List(listOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	return rsList.Items, nil
+}
+
+// GetReplicaSet gets the the daemon set with given name
+func (c *Client) GetReplicaSet(name, namespace string) (*appsv1.ReplicaSet, error) {
+	if err := c.initClient(); err != nil {
+		return nil, err
+	}
+
+	if len(namespace) == 0 {
+		namespace = corev1.NamespaceDefault
+	}
+
+	rs, err := c.apps.ReplicaSets(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return rs, nil
+}
+
+// GetReplicaSetPods returns list of pods for the ReplicaSet
+func (c *Client) GetReplicaSetPods(rs *appsv1.ReplicaSet) ([]corev1.Pod, error) {
+	return common.GetPodsByOwner(c.core, rs.UID, rs.Namespace)
+}
+
+// ValidateReplicaSet checks if the given ReplicaSet is ready within given timeout
+func (c *Client) ValidateReplicaSet(name, namespace string, timeout time.Duration) error {
+	t := func() (interface{}, bool, error) {
+		rs, err := c.GetReplicaSet(name, namespace)
+		if err != nil {
+			return "", true, err
+		}
+
+		if rs.Status.ObservedGeneration == 0 {
+			return "", true, &schederrors.ErrAppNotReady{
+				ID:    name,
+				Cause: "Observed generation is still 0. Check back status after some time",
+			}
+		}
+
+		pods, err := c.GetReplicaSetPods(rs)
```

Comment:
Is retrieving the replicaset pods required here?
ReplicaSet controller updates rs status with pods information so it should be enough to check the status.

Reply:
yep, it's needed since we need the pods to proper print the status of each pod if you check bellow